packages/fluentui/react-bindings/src/hooks/useAutoControlled.ts

```diff
@@ -17,7 +17,18 @@ export const useAutoControlled = <Value>(
   options: UseAutoControlledOptions<Value>,
 ): [Value, React.Dispatch<React.SetStateAction<Value>>] => {
   const { defaultValue, initialValue = undefined, value } = options;
-  const [state, setState] = React.useState<Value>(isUndefined(defaultValue) ? (initialValue as Value) : defaultValue);
+  const [_state, _setState] = React.useState<Value>(isUndefined(defaultValue) ? (initialValue as Value) : defaultValue);
+  const state = isUndefined(value) ? _state : value;
+  const stateRef = React.useRef(state);
 
-  return [isUndefined(value) ? state : value, setState];
+  const setState = React.useCallback((newState: Value) => {
+    if (typeof newState === 'function') {
+      stateRef.current = newState(stateRef.current);
+    } else {
+      stateRef.current = newState;
+    }
+    _setState(stateRef.current);
+  }, []);
+
+  return [state, setState];
```

Comment:
Small update to naming to make it more clear:

```tsx
export const useAutoControlled = <Value>(
  options: UseAutoControlledOptions<Value>,
): [Value, React.Dispatch<React.SetStateAction<Value>>] => {
  const [stateValue, setStateValue] = React.useState<Value>(
    isUndefined(options.defaultValue) ? (options.initialValue as Value) : options.defaultValue,
  );

  const value = isUndefined(options.value) ? stateValue : options.value;
  // Used to avoid dependencies in "setValue"
  const valueRef = React.useRef(value);

  const setValue = React.useCallback((newState: Value) => {
    if (typeof newState === 'function') {
      // Handles functional updates
      // https://reactjs.org/docs/hooks-reference.html#functional-updates
      valueRef.current = newState(valueRef.current);
    } else {
      valueRef.current = newState;
    }

    setStateValue(valueRef.current);
  }, []);

  return [value, setValue];
};
```